Fix smart pointer warnings in FontCoreText.cpp

rniwa opened the pull request:
#### fb455d61079a345fdfbd4f8c5cc61be2f8082c33
<pre>
Fix smart pointer warnings in FontCoreText.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=280857">https://bugs.webkit.org/show_bug.cgi?id=280857</a>

Reviewed by Vitor Roriz.

Deployed more smart pointers to fix clang static analyzer warnings.

* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
(WebCore::FontPlatformData::protectedCustomPlatformData const):
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::createFontWithoutSynthesizableFeatures const):
(WebCore::Font::platformCreateScaledFont const):

Canonical link: <a href="https://commits.webkit.org/284649@main">https://commits.webkit.org/284649@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6ad5d3e7b4b510ea10dec5dffc729fa94962f1da

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/70093 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/49496 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/22856 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/74180 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/21259 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/72210 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/57298 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/21111 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55616 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/14091 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/73159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/45086 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/60450 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/36082 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41746 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17881 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/19628 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63674 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/18228 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75897 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/14318 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/17468 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/63315 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/14358 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/60519 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/63249 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/11271 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4885 "Passed tests") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10709 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/45301 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/60 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/46375 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/47649 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/46116 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

vitorroriz commented on Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
```diff
@@ -516,7 +516,7 @@ RefPtr<Font> Font::createFontWithoutSynthesizableFeatures() const
     CTFontSymbolicTraits fontTraits = CTFontGetSymbolicTraits(getCTFont());
     RetainPtr<CTFontRef> ctFont = createCTFontWithoutSynthesizableFeatures(getCTFont());
-    return createDerivativeFont(ctFont.get(), size, m_platformData.orientation(), fontTraits, m_platformData.syntheticBold(), m_platformData.syntheticOblique(), m_platformData.widthVariant(), m_platformData.textRenderingMode(), m_platformData.customPlatformData());
+    return createDerivativeFont(ctFont.get(), size, m_platformData.orientation(), fontTraits, m_platformData.syntheticBold(), m_platformData.syntheticOblique(), m_platformData.widthVariant(), m_platformData.textRenderingMode(), m_platformData.protectedCustomPlatformData().get());
```
The change itself looks good. However, the `FontCustomPlatformData` raw ptr is used by `createDerivativeFont` for instantiating a `FontPlatformData` object which has a ctor [1] that accepts such raw ptr. This ctor calls yet another `FontPlatformData` ctor [2] that takes a `RefPtr`. Shouldn't we change ctor 1 to get a `RefPtr` directly instead of a RawPtr and then move it?

rniwa replied:
Oh, I didn't realize that. Yes, would do that.

rniwa replied:
Hm... this is annoying. We can't quite do that without including FontCustomPlatformData.h in FontPlatformData.h or else we'd get this build error:
```
/Volumes/Data/safari/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/Ref.h:62:16: error: member access into incomplete type 'const WebCore::FontCustomPlatformData'
   62 |             ptr->deref();
      |                ^
In file included from /Volumes/Data/safari/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource9-mm.mm:1:
In file included from /Volumes/Data/safari/OpenSource/Source/WebCore/WebCorePrefix.h:165:
In file included from /Volumes/Data/safari/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/HashMap.h:25:
In file included from /Volumes/Data/safari/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/HashTable.h:33:
In file included from /Volumes/Data/safari/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/HashTraits.h:26:
In file included from /Volumes/Data/safari/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/HashFunctions.h:26:
/Volumes/Data/safari/OpenSource/WebKitBuild/Debug/usr/local/include/wtf/RefPtr.h:60:47: note: in instantiation of member function 'WTF::DefaultRefDerefTraits<const WebCore::FontCustomPlatformData>::derefIfNotNull' requested here
   60 |     ALWAYS_INLINE ~RefPtr() { RefDerefTraits::derefIfNotNull(PtrTraits::exchange(m_ptr, nullptr)); }
      |                                               ^
In file included from /Volumes/Data/safari/OpenSource/WebKitBuild/Debug/DerivedSources/WebCore/unified-sources/UnifiedSource9-mm.mm:3:
In file included from /Volumes/Data/safari/OpenSource/Source/WebCore/editing/cocoa/AttributedString.mm:31:
In file included from /Volumes/Data/safari/OpenSource/Source/WebCore/platform/graphics/Font.h:26:
/Volumes/Data/safari/OpenSource/Source/WebCore/platform/graphics/FontPlatformData.h:253:327: note: in instantiation of member function 'WTF::RefPtr<const WebCore::FontCustomPlatformData>::~RefPtr' requested here
  253 |     WEBCORE_EXPORT FontPlatformData(RetainPtr<CTFontRef>&&, float size, bool syntheticBold = false, bool syntheticOblique = false, FontOrientation = FontOrientation::Horizontal, FontWidthVariant = FontWidthVariant::RegularWidth, TextRenderingMode = TextRenderingMode::AutoTextRendering, RefPtr<const FontCustomPlatformData> = nullptr);
      |                                                                                                                                                                                                                                                                                                                                       ^
/Volumes/Data/safari/OpenSource/Source/WebCore/platform/graphics/FontPlatformData.h:81:8: note: forward declaration of 'WebCore::FontCustomPlatformData'
   81 | struct FontCustomPlatformData;
      |        ^
1 error generated.
```

rniwa replied:
or so I thought. But `FontCustomPlatformData.h` includes `FontPlatformData.h` so it's more hairy than that. I don't think there is a conceivable way we could do this short of merging those two header files.

vitorroriz replied:
Fair enough, I think the patch itself is fine.

rniwa committed:
Fix smart pointer warnings in FontCoreText.cpp
https://bugs.webkit.org/show_bug.cgi?id=280857

Reviewed by Vitor Roriz.

Deployed more smart pointers to fix clang static analyzer warnings.

* Source/WebCore/platform/graphics/FontCustomPlatformData.h:
(WebCore::FontPlatformData::protectedCustomPlatformData const):
* Source/WebCore/platform/graphics/FontPlatformData.h:
* Source/WebCore/platform/graphics/coretext/FontCoreText.cpp:
(WebCore::Font::createFontWithoutSynthesizableFeatures const):
(WebCore::Font::platformCreateScaledFont const):

Canonical link: https://commits.webkit.org/284649@main
